cdap-app-fabric/src/main/java/co/cask/cdap/app/guice/ServiceStoreModules.java

```diff
@@ -19,69 +19,65 @@
 import co.cask.cdap.data2.dataset2.DefaultDatasetDefinitionRegistry;
+import co.cask.cdap.data2.dataset2.InMemoryDatasetFramework;
 import co.cask.cdap.data2.dataset2.lib.kv.HBaseKVTableDefinition;
 import co.cask.cdap.data2.dataset2.lib.kv.InMemoryKVTableDefinition;
-import co.cask.cdap.data2.dataset2.lib.kv.LevelDBKVTableDefinition;
 import co.cask.cdap.gateway.handlers.DatasetServiceStore;
 import com.google.inject.Module;
 import com.google.inject.PrivateModule;
 import com.google.inject.Scopes;
-import com.google.inject.TypeLiteral;
 import com.google.inject.assistedinject.FactoryModuleBuilder;
+import com.google.inject.multibindings.MapBinder;
 import com.google.inject.name.Names;
 
 /**
  * ServiceStore Guice Modules.
  */
-public class ServiceStoreModules {
+public class ServiceStoreModules extends RuntimeModule {
 
-  public Module getInMemoryModule() {
-    return new PrivateModule() {
-      @Override
-      protected void configure() {
-        install(new FactoryModuleBuilder()
-                  .implement(DatasetDefinitionRegistry.class, DefaultDatasetDefinitionRegistry.class)
-                  .build(DatasetDefinitionRegistryFactory.class));
-
-        bind(new TypeLiteral<DatasetModule>() { }).annotatedWith(Names.named("serviceModule"))
-          .toInstance(new InMemoryKVTableDefinition.Module());
-        bind(ServiceStore.class).to(DatasetServiceStore.class).in(Scopes.SINGLETON);
-        expose(ServiceStore.class);
-      }
-    };
+  @Override
+  public Module getInMemoryModules() {
+    return new ServiceStoreModule(new InMemoryKVTableDefinition.Module());
   }
 
-  public Module getStandaloneModule() {
-    return new PrivateModule() {
-      @Override
-      protected void configure() {
-        install(new FactoryModuleBuilder()
-                  .implement(DatasetDefinitionRegistry.class, DefaultDatasetDefinitionRegistry.class)
-                  .build(DatasetDefinitionRegistryFactory.class));
+  @Override
+  public Module getStandaloneModules() {
+    return new ServiceStoreModule(new InMemoryKVTableDefinition.Module());
+  }
 
-        bind(new TypeLiteral<DatasetModule>() { }).annotatedWith(Names.named("serviceModule"))
-          .toInstance(new LevelDBKVTableDefinition.Module());
-        bind(ServiceStore.class).to(DatasetServiceStore.class).in(Scopes.SINGLETON);
-        expose(ServiceStore.class);
-      }
-    };
+  @Override
+  public Module getDistributedModules() {
+    return new ServiceStoreModule(new HBaseKVTableDefinition.Module());
   }
 
-  public Module getDistributedModule() {
-    return new PrivateModule() {
-      @Override
-      protected void configure() {
-        install(new FactoryModuleBuilder()
-                  .implement(DatasetDefinitionRegistry.class, DefaultDatasetDefinitionRegistry.class)
-                  .build(DatasetDefinitionRegistryFactory.class));
+  private static final class ServiceStoreModule extends PrivateModule {
+
+    private final DatasetModule tableModule;
```

Comment:
I only thought so, because we only use modules of ``KVTable`` in ``ServiceStore``, right?